usr/usr.sbin/catman/catman.8
.\" Copyright (c) 1980 Regents of the University of California.
.\" All rights reserved.  The Berkeley software License Agreement
.\" specifies the terms and conditions for redistribution.
.\"
.\"	@(#)catman.8	6.5.2 (2.11BSD) 1996/11/16
.\"
.TH CATMAN 8 "October 23, 1996"
.UC 4
.SH NAME
catman \- create the cat files for the manual
.SH SYNOPSIS
.B /usr/sbin/catman
[
.B \-p
] [
.B \-n
] [
.B \-w
] [
.B \-M
.I path
]
[ sections ]
.SH DESCRIPTION
.I Catman
creates the preformatted versions of the on-line manual from the nroff
input files.
Each manual page is examined and those whose preformatted versions are
missing or out of date are recreated.
If any changes are made,
.I catman
will recreate the
.B whatis
database.
.PP
If there is one parameter not starting with a `\-',
it is taken to be a list of manual sections to look in.
For example
.IP
.B catman 123
.LP
will cause the updating to only happen to manual sections
1, 2, and 3.
.PP
Options:
.TP
.B \-n
prevents creations of the
.B whatis
database.
.TP
.B \-p
prints what would be done instead of doing it.
.TP
.B \-w
causes only the 
.B whatis
database to be created.
No manual reformatting is done.
.TP
.B \-M
updates manual pages located in the set of directories specified by
.I path
(/usr/man by default).
.I Path
has the form of a colon (`:') separated list of directory names,
for example `/usr/local/man:/usr/man'.
If the environment variable `MANPATH' is set,
its value is used for the default path.
.PP
If the nroff source file contains only a line of the form `.so manx/yyy.x',
a symbolic link is made in the catx directory to the appropriate
preformatted manual page.
This feature allows easy distribution of the preformatted manual pages
among a group of associated machines with
.I rdist(1).
The nroff sources need not be distributed to all machines,
thus saving the associated disk space.
As an example, consider a local network with 5 machines,
called mach1 through mach5.
Suppose mach3 has the manual page nroff sources.
Every night, mach3 runs
.I catman
via
.I cron(8)
and later runs
.I rdist
with a distfile that looks like:
.IP
.nf
.ta \w'${MANUAL'u

MANSLAVES = ( mach1 mach2 mach4 mach5 )

MANUALS = (/usr/man/cat[1-8no] /usr/man/whatis)

${MANUALS} -> ${MANSLAVES}
	install -R;
	notify root;

.fi
.SH FILES
.nf
.ta \w'/usr/sbin/makewhatis   'u
/usr/man	default manual directory location
/usr/man/man?/*.*	raw (nroff input) manual sections
/usr/man/cat?/*.*	preformatted manual pages
/usr/man/whatis	whatis database
/usr/sbin/makewhatis	command script to make whatis database
.fi
.DT
.SH "SEE ALSO"
man(1), cron(8), rdist(1)
.SH BUGS
Acts oddly on nights with full moons.
.PP
The need for catman(8) is almost but not quite gone.  Most of the manpages
have been moved out of /usr/src/man into the sourcecode hierarchy.
The recreation of the whatis database is the main use of catman now.


usr/usr.sbin/catman/catman.c
/*
 * Copyright (c) 1980 Regents of the University of California.
 * All rights reserved.  The Berkeley software License Agreement
 * specifies the terms and conditions for redistribution.
 */

#if	!defined(lint) && defined(DOSCCS)
char copyright[] =
"@(#) Copyright (c) 1980 Regents of the University of California.\n\
 All rights reserved.\n";

static char sccsid[] = "@(#)catman.c	5.7.1 (2.11BSD) 1996/10/21";
#endif

/*
 * catman: update cat'able versions of manual pages
 *  (whatis database also)
 */
#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/file.h>
#include <sys/time.h>
#include <sys/dir.h>
#include <ctype.h>
#include <string.h>
#include <stdlib.h>

char	buf[BUFSIZ];
char	pflag;
char	nflag;
char	wflag;
char	man[MAXNAMLEN+6] = "manx/";
int	exstat = 0;
char	cat[MAXNAMLEN+6] = "catx/";
char	lncat[MAXNAMLEN+9] = "../catx/";
char	*manpath = "/usr/man";
char	*sections = "12345678ln";
char	*makewhatis = "/usr/sbin/makewhatis";

main(ac, av)
	int ac;
	char *av[];
{
	char *mp, *nextp;

	if ((mp = getenv("MANPATH")) != NULL)
		manpath = mp;

	ac--, av++;
	while (ac > 0 && av[0][0] == '-') {
		switch (av[0][1]) {

		case 'p':
			pflag++;
			break;

		case 'n':
			nflag++;
			break;

		case 'w':
			wflag++;
			break;

		case 'M':
		case 'P':
			if (ac < 1) {
				fprintf(stderr, "%s: missing path\n",
				    av[0]);
				exit(1);
			}
			ac--, av++;
			manpath = *av;
			break;

		default:
			goto usage;
		}
		ac--, av++;
	}
	if (ac > 1) {
usage:
		printf("usage: catman [ -p ] [ -n ] [ -w ] [ -M path ] [ sections ]\n");
		exit(-1);
	}
	if (ac == 1)
		sections = *av;
	for (mp = manpath; mp && ((nextp = index(mp, ':')), 1); mp = nextp) {
		if (nextp)
			*nextp++ = '\0';
		doit(mp);
	}
	exit(exstat);
}

doit(mandir)
	char *mandir;
{
	register char *msp, *csp, *sp;
	int changed = 1;
	int status;

	if (wflag)
		goto whatis;
	if (chdir(mandir) < 0) {
		sprintf(buf, "catman: %s", mandir);
		perror(buf);
		/* exstat = 1; */
		return;
	}
	if (pflag)
		printf("cd %s\n", mandir);
	msp = &man[5];
	csp = &cat[5];
	(void) umask(0);
	for (sp = sections; *sp; sp++) {
		register DIR *mdir;
		register struct direct *dir;
		struct stat sbuf;

		man[3] = cat[3] = *sp;
		*msp = *csp = '\0';
		if ((mdir = opendir(man)) == NULL) {
			/* sprintf(buf, "catman: opendir: %s", man); */
			/* perror(buf); */
			/* exstat = 1; */
			continue;
		}
		if (stat(cat, &sbuf) < 0) {
			register char *cp;

			(void) strcpy(buf, cat);
			cp = rindex(buf, '/');
			if (cp && cp[1] == '\0')
				*cp = '\0';
			if (pflag)
				printf("mkdir %s\n", buf);
			else if (mkdir(buf, 0777) < 0) {
				sprintf(buf, "catman: mkdir: %s", cat);
				perror(buf);
				exstat = 1;
				continue;
			}
			(void) stat(cat, &sbuf);
		}
		if (access(cat, R_OK|W_OK|X_OK) == -1) {
			sprintf(buf, "catman: %s", cat);
			perror(buf);
			exstat = 1;
			continue;
		}
		if ((sbuf.st_mode & S_IFMT) != S_IFDIR) {
			fprintf(stderr, "catman: %s: Not a directory\n", cat);
			exstat = 1;
			continue;
		}
		while ((dir = readdir(mdir)) != NULL) {
			time_t time;
			register char *tsp;
			FILE *inf;
			int  makelink;

			if (dir->d_ino == 0 || dir->d_name[0] == '.')
				continue;
			/*
			 * Make sure this is a man file, i.e., that it
			 * ends in .[0-9] or .[0-9][a-z]
			 */
			tsp = rindex(dir->d_name, '.');
			if (tsp == NULL)
				continue;
			if (!isdigit(*++tsp) && *tsp != *sp)
				continue;
			if (*++tsp && !isalpha(*tsp))
				continue;
			if (*tsp && *++tsp)
				continue;
			(void) strcpy(msp, dir->d_name);
			if ((inf = fopen(man, "r")) == NULL) {
				sprintf(buf, "catman: %s");
				perror(buf);
				exstat = 1;
				continue;
			}
			makelink = 0;
			if (getc(inf) == '.' && getc(inf) == 's'
			    && getc(inf) == 'o') {
				if (getc(inf) != ' ' ||
				    fgets(lncat+3, sizeof(lncat)-3, inf)==NULL) {
					fclose(inf);
					continue;
				}
				if (lncat[strlen(lncat)-1] == '\n')
					lncat[strlen(lncat)-1] = '\0';
				if (strncmp(lncat+3, "man", 3) != 0) {
					fclose(inf);
					continue;
				}
				bcopy("../cat", lncat, sizeof("../cat")-1);
				makelink = 1;
			}
			fclose(inf);
			(void) strcpy(csp, dir->d_name);
			if (stat(cat, &sbuf) >= 0) {
				time = sbuf.st_mtime;
				(void) stat(man, &sbuf);
				if (time >= sbuf.st_mtime)
					continue;
				(void) unlink(cat);
			}
			if (makelink) {
				/*
				 * Don't unlink a directory by accident.
				 */
				if (stat(lncat+3, &sbuf) >= 0 &&
				    (((sbuf.st_mode&S_IFMT)==S_IFREG) ||
				     ((sbuf.st_mode&S_IFMT)==S_IFLNK)))
					(void) unlink(cat);
				if (pflag)
					printf("ln -s %s %s\n", lncat, cat);
				else
					if (symlink(lncat, cat) == -1) {
						sprintf(buf, "catman: symlink: %s", cat);
						perror(buf);
						exstat = 1;
						continue;
					}
			}
			else {
				sprintf(buf, "nroff -man %s > %s", man, cat);
				if (pflag)
					printf("%s\n", buf);
				else if ((status = system(buf)) != 0) {
					fprintf(stderr, "catman: nroff: %s: exit status %d: Owooooo!\n", cat, status);
					exstat = 1;
					continue;
				}
			}
			changed = 1;
		}
		closedir(mdir);
	}
	if (changed && !nflag) {
whatis:
		sprintf(buf, "%s %s", makewhatis, mandir);
		if (pflag)
			printf("%s\n", buf);
		else if ((status = system(buf)) != 0) {
			fprintf(stderr, "catman: %s: exit status %d\n",
			    buf, status);
			exstat = 1;
		}
	}
	return;
}
